services/src/main/java/org/keycloak/services/clientpolicy/executor/AutomaticClientRegistrationExecutor.java

```diff
@@ -0,0 +1,371 @@
+
+import org.keycloak.common.util.Base64Url;
+import org.keycloak.constants.AdapterConstants;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.models.Constants;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.OAuth2Constants;
+import org.keycloak.OAuthErrorException;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.protocol.oidc.grants.ciba.CibaGrantType;
+import org.keycloak.representations.idm.ClientPolicyExecutorConfigurationRepresentation;
+import org.keycloak.representations.openid_federation.EntityStatement;
+import org.keycloak.representations.openid_federation.RPMetadata;
+import org.keycloak.services.clientpolicy.ClientPolicyContext;
+import org.keycloak.services.clientpolicy.ClientPolicyException;
+import org.keycloak.services.clientpolicy.context.PreAuthorizationRequestContext;
+import org.keycloak.services.clientregistration.AbstractClientRegistrationProvider;
+import org.keycloak.services.clientregistration.ClientRegistrationAuth;
+import org.keycloak.util.JsonSerialization;
+
+public class AutomaticClientRegistrationExecutor extends AbstractClientRegistrationProvider 
+                                implements ClientPolicyExecutorProvider<ClientPolicyExecutorConfigurationRepresentation> {
+
+    private static final Logger logger = Logger.getLogger(AutomaticClientRegistrationExecutor.class);
+
+    public AutomaticClientRegistrationExecutor(KeycloakSession session) {
+        super(session);
+        EventBuilder event = new EventBuilder(session.getContext().getRealm(), session);
+        super.setEvent(event);
+        super.setAuth(new ClientRegistrationAuth(session,this,event,"openid-connect"));
+    }
+
+    @Override
+    public void executeOnEvent(ClientPolicyContext context) throws ClientPolicyException {
+
+        PreAuthorizationRequestContext paContext = null;
+        switch (context.getEvent()) {
+            case PRE_AUTHORIZATION_REQUEST:
+                paContext = (PreAuthorizationRequestContext) context;
+                String requestParam = paContext.getRequestParameters().getFirst(OIDCLoginProtocol.REQUEST_PARAM);
+
+                String [] splits = requestParam.split("\\.");
+
+                byte[] serializedRequestObject = Base64Url.decode(splits[1]);
+                AuthorizationEndpointRequestObjectForAutomaticClientRegistration oidcRequest = null;
+                try {
+                    oidcRequest = JsonSerialization.readValue(serializedRequestObject, AuthorizationEndpointRequestObjectForAutomaticClientRegistration.class);
+                } catch (Exception e) {
+                    logger.debugv("exception on parsing request object");
+                    throw new ClientPolicyException(OAuthErrorException.INVALID_REQUEST_OBJECT, e.getMessage());
+                }
+
+                // getting  RP metadata with which client is created.
+                RPMetadata rpMetadata = oidcRequest.getMetadata().getRelyingPartyMetadata();
```

Comment:
Based on [[12.1.1.1. ](https://openid.net/specs/openid-federation-1_0.html#section-12.1.1.1)[Using a Request Object](https://openid.net/specs/openid-federation-1_0.html#name-using-a-request-object)](https://openid.net/specs/openid-federation-1_0.html#name-using-a-request-object) automatic client registration is extended 3.1.2 in [OpenID Connect Core 1.0](https://openid.net/specs/openid-federation-1_0.html#OpenID.Core) [[OpenID.Core](https://openid.net/specs/openid-federation-1_0.html#OpenID.Core)], not the EntityStatement.
So, the saved client must be based on _entityStatementInRequestObject_ from AutomaticClientRegistrationCondition (Trust Chain Resolution)